src/Open3D/Visualization/Rendering/Filament/FilamentRenderToBuffer.cpp

```diff
@@ -0,0 +1,179 @@
+#include <filament/Renderer.h>
+#include <filament/Scene.h>
+#include <filament/SwapChain.h>
+#include <filament/Texture.h>
+#include <filament/View.h>
+#include <filament/Viewport.h>
+
+#include "FilamentEngine.h"
+#include "FilamentRenderer.h"
+#include "FilamentScene.h"
+#include "FilamentView.h"
+#include "Open3D/Utility/Console.h"
+
+namespace open3d {
+namespace visualization {
+
+FilamentRenderToBuffer::FilamentRenderToBuffer(filament::Engine& engine)
+    : engine_(engine) {
+    renderer_ = engine_.createRenderer();
+}
+
+FilamentRenderToBuffer::FilamentRenderToBuffer(filament::Engine& engine,
+                                               FilamentRenderer& parent)
+    : parent_(&parent), engine_(engine) {
+    renderer_ = engine_.createRenderer();
+}
+
+FilamentRenderToBuffer::~FilamentRenderToBuffer() {
+    engine_.destroy(swapchain_);
+    engine_.destroy(renderer_);
+
+    if (buffer_) {
+        free(buffer_);
+        buffer_ = nullptr;
+
+        bufferSize_ = 0;
+    }
+
+    if (parent_) {
+        parent_->OnBufferRenderDestroyed(this);
+        parent_ = nullptr;
+    }
+}
+
+void FilamentRenderToBuffer::SetDimensions(const std::size_t width,
+                                           const std::size_t height) {
+    if (swapchain_) {
+        engine_.destroy(swapchain_);
+    }
+
+    swapchain_ = engine_.createSwapChain(width, height,
+                                         filament::SwapChain::CONFIG_READABLE);
+    view_->SetViewport(0, 0, width, height);
+
+    width_ = width;
+    height_ = height;
+
+    bufferSize_ = width * height * 3 * sizeof(std::uint8_t);
+    if (buffer_) {
+        buffer_ = static_cast<std::uint8_t*>(realloc(buffer_, bufferSize_));
+    }
+}
+
+void FilamentRenderToBuffer::CopySettings(const View* view) {
+    auto* downcast = dynamic_cast<const FilamentView*>(view);
+    // NOTE: This class used to copy parameters from the view into a view
+    // managed by this class. However, the copied view caused anomalies when
+    // rendering an image for export. As a workaround, we keep a pointer to the
+    // original view here instead.
+    view_ = const_cast<FilamentView*>(downcast);
+    if (downcast) {
+        auto vp = view_->GetNativeView()->getViewport();
+        SetDimensions(vp.width, vp.height);
+    }
+}
+
+View& FilamentRenderToBuffer::GetView() { return *view_; }
+
+using PBDParams = std::tuple<FilamentRenderToBuffer*,
+                             FilamentRenderToBuffer::BufferReadyCallback>;
+void FilamentRenderToBuffer::RequestFrame(Scene* scene,
+                                          BufferReadyCallback callback) {
+    if (!scene) {
+        utility::LogDebug(
+                "No Scene object was provided for rendering into buffer");
+        callback({0, 0, nullptr, 0});
+        return;
+    }
+
+    if (pending_) {
+        utility::LogWarning(
+                "Render to buffer can process only one request at time");
+        callback({0, 0, nullptr, 0});
```

Comment:
I guess it depends on your view of error handling. I (and the original author, I think) consider it rude to abort, especially as a library. Maybe the code is processing a number of scenes? Why not do what you can, rather than aborting at the first sign of a problem? Also, on macOS and Windows, the user probably won't see the console warning, so we probably need to rethink how we handle errors in general.